### DISCUSSION TITLE
In Pivot with multiple `ON` columns, only show combinations that exist

### DISCUSSION THREAD

**Discussion by Alex-Monahan** [duckdb/duckdb#6754] [Ideas]
Howdy! This may not be feasible, depending on the implementation, but I figured I would at least bring it up as an idea! 

When creating a pivot in DuckDB with multiple columns specified in the `ON` clause, the cartesian product of column name combinations are produced. Spreadsheet software will only show the column combinations that are present in the underlying data. 

Would it be possible for DuckDB to also only show the column combinations that exist in the data?

```sql
CREATE TABLE Cities(Country VARCHAR, Name VARCHAR, Year INT, Population INT);
INSERT INTO Cities VALUES ('NL', 'Amsterdam', 2000, 1005);
INSERT INTO Cities VALUES ('NL', 'Amsterdam', 2010, 1065);
INSERT INTO Cities VALUES ('NL', 'Amsterdam', 2020, 1158);
INSERT INTO Cities VALUES ('US', 'Seattle', 2000, 564);
INSERT INTO Cities VALUES ('US', 'Seattle', 2010, 608);
INSERT INTO Cities VALUES ('US', 'Seattle', 2020, 738);
INSERT INTO Cities VALUES ('US', 'New York City', 2000, 8015);
INSERT INTO Cities VALUES ('US', 'New York City', 2010, 8175);
INSERT INTO Cities VALUES ('US', 'New York City', 2020, 8772);
```

```sql
PIVOT Cities ON Country, Name USING SUM(Population) GROUP BY Year;
```

| Year | NL_Amsterdam | NL_New York City | NL_Seattle | US_Amsterdam | US_New York City | US_Seattle |
|------|--------------|------------------|------------|--------------|------------------|------------|
| 2000 | 1005         |                  |            |              | 8015             | 564        |
| 2010 | 1065         |                  |            |              | 8175             | 608        |
| 2020 | 1158         |                  |            |              | 8772             | 738        |

Excel as a comparison:
![image](https://user-images.githubusercontent.com/52226177/225666500-8d02756c-6d4b-46e5-9ba8-1279eaf7d14f.png)

One workaround would be to pre-concatenate the columns prior to pivoting:
```sql
PIVOT (SELECT *, Country || '_' || Name as Country_Name FROM Cities) ON Country_Name USING SUM(Population) GROUP BY Year;
```

| Year | NL_Amsterdam | US_New York City | US_Seattle |
|------|--------------|------------------|------------|
| 2000 | 1005         | 8015             | 564        |
| 2010 | 1065         | 8175             | 608        |
| 2020 | 1158         | 8772             | 738        |

I tried to do the concatenation within the `PIVOT` statement, but it does not seem to support arbitrary expressions:
```sql
PIVOT Cities ON Country || '_' || Name USING SUM(Population) GROUP BY Year;
Error: Parser Error: syntax error at or near "||"
LINE 1: PIVOT Cities ON Country || '_' || Name USING SUM(Population) GR...
                                ^

PIVOT Cities ON (Country || '_' || Name) USING SUM(Population) GROUP BY Year;
Error: Parser Error: syntax error at or near "||"
LINE 1: PIVOT Cities ON (Country || '_' || Name) USING SUM(Population) G...
                                 ^
```

In Pandas, there is a parameter that controls this behavior: `dropna`. If true, entirely empty columns are removed.
https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

**Comment by Mytherin** [comment]
Good point! There is no real reason we can't pivot on expressions - added this in  #6758

**Reply by Alex-Monahan** [reply]
Phenomenal! I think I'll use the `|| '_' ||` approach at the day job. Thank you again!

**Comment by gfgw** [comment]
I just ran into the same issue. The number of columns is exploding because of this. Most of the columns are meaningless, making the result unusable.

**Reply by Mytherin** [reply]
As mentioned in the discussion above, you can use `CONCAT` if you want to only pivot on combinations that exist in the data, e.g.:

```sql
 PIVOT Cities ON CONCAT(Country, '_', Name);
```
